[LangRef] Clarify pointer capture spec (#194647)
llvm/llvm-project: llvm/docs/LangRef.rst

    rules <pointeraliasing>`. We further distinguish whether only read accesses
    are allowed, or both reads and writes.

These two cases are discussed in more detail in the following.

Provenance capture
^^^^^^^^^^^^^^^^^^

> If an argument does not capture the provenance the pointer, accesses that are

Reviewer: ```suggestion
If an argument does not capture the provenance of the pointer, accesses that are
```

based on the argument and are performed after the function returns (or unwinds)
cause undefined behavior. In the case where only read provenance is captured,
only stores cause undefined behavior.

> More formally, an argument that does not capture provenance will return a new

Reviewer: Not sure return reads best here, how about:
```suggestion
More formally, an argument that does not capture provenance effectively provides to the callee a new
```

pointer with the same address and a derived provenance, where the derived
provenance initially has the same permissions as the original, but all access
permissions are removed once the function returns (or unwinds). Or in the case
of only capturing read provenance, the permissions are changed to only allow
read accesses.

This means that the following code is well-defined in isolation:

.. code-block:: llvm

    define void @f(ptr captures(address) %a, ptr %b) {
      store ptr %a, ptr %b
      ret void
    }

Even though the pointer is stored in another location that persists past the
return of the function, this does not cause undefined behavior in itself. It
depends on whether/how the pointer will be used in the future.

.. code-block:: llvm

    call void @f(ptr captures(address) %a, ptr %b)
    %a2 = load ptr, ptr %b ; This is still well-defined

>     load i64, ptr %a2 ; This causes undefined behavior

Reviewer: Would it make sense to add another example, showing this load would be fine with `captures(address, read_provenance)` on `%a` (but the store to the loaded ptr would be UB)?

Author: Done

In this example, the persisted pointer is accessed after the return of the

> function, which causes undefined behavior.

Reviewer: I would find it helpful to mention that, e.g., if an argument is annotated with `captures(address)`, this is essentially a promise that no dereference copies are done after the return (or otherwise UB), so the optimizer is allowed to make more aggressive assumptions (which I believe, among other things, would help understand why the spec is now defined as a semantic contract / in terms of UB); though not strong on this.

Author: Isn't this covered by the first paragraph in this section?

Address capture
^^^^^^^^^^^^^^^

The address of the pointer is considered to be captured if the function can

> exhibit different behavior based on the address of the pointer.

Reviewer: Different observable behavior or is this the part we are unsure how to formalize this properly?

Author: After thinking about this a bit more, I think my ideal here would be that even the observable behavior of the function may change, and the attribute is just a statement that such changes in behavior are allowed -- in a similar way that you might see comparisons between lifetime-disjoint stack allocations evaluate to either true or false, and both behaviors are allowed.

One of the main things that address-capture controls are things like call slot optimization, and I'd like `captures(none)` to be well-defined even in cases where the call can in fact observe whether or not an intermediate allocation has been elided, but we want to convey that both options are legal.

But I'm not sure how to best specify that.

This definition is chosen to allow capture analysis to continue with the return
value in the usual fashion.

The following describes possible ways to capture a pointer in more detail,
where unqualified uses of the word "capture" refer to capturing both address
and provenance.

> 1. The call stores any bit of the pointer carrying information into a place,

Reviewer: Hmm, cataloguing the ways a pointer could escape could still be helpful? (I'm referring to example 1 in particular, which could be still maintained IIUC, though not sure how it would fit best with the new capture spec).

Author: Yeah, I agree that this is still somewhat useful. I think we need to distinguish between the spec of `captures` (where a store of the pointer is fine by itself) from the inference behavior, where we have to conservatively assume that any store of the pointer captures, as we don't know what will be done with it outside the function. I guess I could add an extra section on this.

Author: I've added a section on inference now.

      store ptr null, ptr @glb2

      store ptr %c,   ptr @glb3
      call void @g() ; If @g makes a copy of %c that outlives this call (@f), %c is captured
      store ptr null, ptr @glb3

      %i = ptrtoint ptr %d to i64
      %j = trunc i64 %i to i32
      store i32 %j, ptr @glbi ; %d is captured

      ret ptr %e ; %e is captured
    }

2. The call stores any bit of the pointer carrying information into a place,
   and the stored bits can be safely read from the place by another thread via
   synchronization.

.. code-block:: llvm

    @lock = global i1 true

    define void @f(ptr %a) {
      store ptr %a, ptr @glb
      store atomic i1 false, ptr @lock release ; %a is captured because another thread can safely read @glb
      store ptr null, ptr @glb
      ret void
    }

3. The call's behavior depends on any bit of the pointer carrying information
   (address capture only).

.. code-block:: llvm

    @glb = global i8 0

    define void @f(ptr %a) {
      %c = icmp eq ptr %a, @glb
      br i1 %c, label %BB_EXIT, label %BB_CONTINUE ; captures address of %a only
    BB_EXIT:
      call void @exit()
      unreachable
    BB_CONTINUE:
      ret void
    }

> 4. The pointer is used as the pointer operand of a volatile access.

Reviewer: I have some doubts on this, but, in any case, this would now not be mentioned anymore anywhere else?

Reviewer: (To be precise: I do think volatile accesses are a separate concern, unrelated to pointer capture, but would like to be sure we are removing it because of this.)

Author: This just got dropped because I dropped the listing.

I think the connection to volatile is that a volatile access observes the address of the pointer. CaptureTracking currently also captures the provenance, but I think this is unnecessary (this is just what I conservatively did when implementing the address/provenance split).

Author: I've added the mention of volatile accesses to the inference section.

This definition is chosen to allow capture analysis to continue with the return
value in the usual fashion.

Inference
^^^^^^^^^

The definitions given above are permissive to facilitate frontend-driven
annotations based on language semantics. It is useful to also consider in
which cases we can infer lack of captures based on conservative function-local

> analysis. In that case, the following rules hold:

Reviewer 2: I would restate it as "here is an example of a set of rules a transform could use to infer nocapture", to make it clear this isn't actually supposed to be defining the semantics of anything..
